File: Sample_Study/ELT_Study.py
import pandas as pd
from sqlalchemy import create_engine

source_engine = create_engine('postgresql://user:pass@source_host/db')
target_engine = create_engine('postgresql://user:pass@target_host/dw')

# 1. Extract: 소스 데이터 추출
df = pd.read_sql("SELECT id, name, amount FROM orders", source_engine)

# 2. Load: 가공하지 않고 타겟 저장소의 임시 테이블(Staging Table)에 그대로 적재
df.to_sql('stg_orders', target_engine, if_exists='replace', index=False)

# 3. Transform: 타겟 데이터 웨어하우스의 연산 능력을 활용해 SQL로 데이터 변환 및 최종 적재
with target_engine.begin() as conn:
    transform_query = """
        INSERT INTO fact_orders (id, name, amount)
        SELECT id, UPPER(name), amount 
        FROM stg_orders;
    """
    conn.execute(transform_query)

File: Sample_Study/ELT_Study.java
import java.sql.*;

public class ELT_Study {
    public static void main(String[] args) throws Exception {
        String sourceUrl = "jdbc:postgresql://source_host/db";
        String targetUrl = "jdbc:postgresql://target_host/dw";

        // 1 & 2. Extract & Load: 소스에서 읽어와 변환 없이 타겟의 Staging 테이블로 바로 스트리밍 적재
        try (Connection srcConn = DriverManager.getConnection(sourceUrl, "user", "pass");
             Connection tgtConn = DriverManager.getConnection(targetUrl, "user", "pass");
             Statement srcStmt = srcConn.createStatement();
             ResultSet rs = srcStmt.executeQuery("SELECT id, name, amount FROM orders");
             PreparedStatement tgtPstmt = tgtConn.prepareStatement("INSERT INTO stg_orders (id, name, amount) VALUES (?, ?, ?)")) {
            
            while (rs.next()) {
                tgtPstmt.setInt(1, rs.getInt("id"));
                tgtPstmt.setString(2, rs.getString("name")); // 변환 없이 그대로 입력
                tgtPstmt.setDouble(3, rs.getAmount());
                tgtPstmt.addBatch();
            }
            tgtPstmt.executeBatch();
        }

        // 3. Transform: 타겟 데이터 웨어하우스 내부에서 SQL(UPPER 함수)을 통해 변환 후 최종 테이블로 이동
        try (Connection tgtConn = DriverManager.getConnection(targetUrl, "user", "pass");
             Statement tgtStmt = tgtConn.createStatement()) {
            
            String transformSql = 
                "INSERT INTO fact_orders (id, name, amount) " +
                "SELECT id, UPPER(name), amount FROM stg_orders";
                
            tgtStmt.executeUpdate(transformSql);
        }
    }
}
